### PR TITLE
TINY-4058: Fix cursors being placed next to pre elements

### PR DESCRIPTION
**Pull request by ashfordneil**
Related Ticket: TINY-4058

Description of Changes:
There were bugs (exceptions being thrown and some content loss) being caused by pressing enter while the cursor was immediately before a `pre` element. I found two ways that the cursor ended up at this invalid location, and prevented them from occurring.
* Pressing enter with the cursor (`|`) like this: `<pre>|Hello</pre>` while `br_in_pre` is set to false in the editor settings
* Calling `editor.selection.setCursorLocation()`

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* ~[ ] Branch prefixed with `feature/` for new features (if applicable)~
* ~[ ] License headers added on new files (if applicable)~

Review:
* [x] Milestone set
* [x] Review comments resolved

Other issues:
* [x] This is currently leading to a different failing test. I'm looking for feedback on how to approach that scenario.

GitHub issues (if applicable): N/A

### CODE REVIEW

File: modules/tinymce/src/core/main/ts/api/html/Schema.ts [hunk 1]
```diff
@@ -478,7 +478,8 @@ function Schema(settings?: SchemaSettings): Schema {
     'noshade nowrap readonly selected autoplay loop controls');
   const nonEmptyElementsMap = createLookupTable('non_empty_elements', 'td th iframe video audio object ' +
     'script pre code', shortEndedElementsMap);
-  const moveCaretBeforeOnEnterElementsMap = createLookupTable('move_caret_before_on_enter_elements', 'table', nonEmptyElementsMap);
+  const moveCaretBeforeOnEnterElementsMap = createLookupTable('move_caret_before_on_enter_elements', 'td th igrame video audio object ' +
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/html/Schema.ts:481]: `igrame` isn't an element as far as I know? :stuck_out_tongue:

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/html/Schema.ts:481]: Also to remove some duplication, should we create something like a `base` list and use it for both this and `nonEmptyElementsMap`?

**Review by lnewson — APPROVED** [review]

**Review by ltrouton — APPROVED** [review]

File: modules/tinymce/changelog.txt [hunk 2]
```diff
@@ -21,6 +21,8 @@ Version 5.5.0 (TBD)
     Fixed the `DOMUtils.split()` API incorrectly removing some content #TINY-6294
     Fixed pressing escape not focusing the editor when using multiple toolbars #TINY-6230
     Fixed the `dirty` flag not being correctly set during an `AddUndo` event #TINY-4707
+    Fixed `editor.selection.setCursorLocation` moving the cursor immediately outside pre elements in some circumstances #TINY-4058
```
Comment (metricjs) [inline modules/tinymce/changelog.txt:24]: "immediately outside" - just wondering if a customer will understand that. @tylerkelly13 ?

File: modules/tinymce/changelog.txt [hunk 3]
```diff
@@ -21,6 +21,8 @@ Version 5.5.0 (TBD)
...
+    Fixed the `br_in_pre` setting (when false) throwing an exception when pressing enter inside pre elements #TINY-4058
```
Comment (metricjs) [inline modules/tinymce/changelog.txt:25]: I wonder if we should move the "when false" outside of brackets somehow... It is an awkward sentence to structure without the brackets though

**Review by metricjs — APPROVED** [review]

File: modules/tinymce/changelog.txt (hunk 2 above)
Comment (tylerkelly13) [reply modules/tinymce/changelog.txt:24]: I'm not sure what that means... do you mean `editor.selection.setCursorLocation` was placing the cursor in-front of the `pre` block when instructed to place the cursor on content within the pre block?

Comment (ashfordneil) [reply modules/tinymce/changelog.txt:24]: In this case, I found the bug by calling `setCursorLocation()` without any arguments (intended to place the cursor at the start of the document). But yes, it was meant to put the cursor on the content within the pre block, and instead placed it before the block.

File: modules/tinymce/changelog.txt (hunk 3 above)
Comment (ashfordneil) [reply modules/tinymce/changelog.txt:25]:
```
Fixed an exception being thrown when pressing enter inside pre elements while `br_in_pre` setting is false
```
How does that sound?

**Pull request merged**